### PR TITLE
[wizlighting] Wiz Lighting Binding initial contribution

### PR DESCRIPTION
**Pull request by basriram**

### CODE REVIEW

**Comment by martinvw** [comment]
Thanks for your PR! 

Please include the build in the parent `pom.xml` in that case it will already be tested and analysed by travis. And it will become available in [jfrog builds list](https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/binding/).

Please also check:

http://docs.openhab.org/developers/development/bindings.html
http://docs.openhab.org/developers/development/guidelines.html

**Comment by wborn** [comment]
Hey @basriram does the binding support other form factors than just bulbs? If it does, I think it would be better to replace all references to bulb(s) with light(s) or just omit the reference to the this specific form factor of a light. E.g. "Wiz Lighting binding". :-) 

Wiz themselves also do this on their [homepage](https://www.wiz.world/en/), e.g. "Connect your lights to the Internet of Things".

**Comment by basriram** [comment]
Makes sense. I will remove references to bulb from the docs.

File: addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        logger.debug("Result {}", urlBuilder.toString());
+        Request request = httpClient.newRequest(urlBuilder.toString()).header("Authorization", "Bearer " + accessToken)
+                .timeout(10, TimeUnit.SECONDS);
+        ContentResponse response = request.send();
+        String result = response.getContentAsString();
+        logger.debug("Invite Result {}", result);
+        return result.contains("\"success\":true");
+    }
+
+    protected int retriveHomeId(String accessToken) throws InterruptedException, TimeoutException, ExecutionException {
```
Comment (triller-telekom) [inline addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java]: `retriveHomeId` -> `retrieveHomeId`

File: addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        logger.debug("User Details Result {}", result);
+        UserDTO data = gson.fromJson(response.getContentAsString(), UserDTO.class);
+        if (data != null) {
+            homeId = data.getHomeId();
+        }
+
+        return homeId;
+    }
+
+    protected HomeDTO.Light[] retriveHomeLights(String accessToken, int homeId)
```
Comment (triller-telekom) [inline addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java]: `retriveHomeLights` -> `retrieveHomeLights`

File: addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            throws InterruptedException, TimeoutException, ExecutionException {
+        AuthenticationStore a = httpClient.getAuthenticationStore();
```
Comment (triller-telekom) [inline addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java]: The next 8 lines are the same as in `retrieveHomeId`, please extract them into a method and provide the URL as a parameter and return the `ContentResponse`.

File: addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/entities/Param.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ */
+package org.openhab.binding.wizlighting.internal.entities;
+
+/**
+ * This POJO represents an abstract Request Param
+ *
+ * @author Sriram Balakrishnan - Initial contribution
+ *
+ */
+public abstract class Param {
```
Comment (triller-telekom) [inline addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/entities/Param.java]: If this class in empty, i.e. with no implementation code, would it not be sufficient to change it to an `interface` instead?

File: addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/utils/NetworkUtils.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
+/**
```
Comment (triller-telekom) [inline addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/utils/NetworkUtils.java]: This class should not be part of a binding but rather be provided by the underlying framework: Eclipse SmartHome.

Please create a PR against https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java

To obtain a specific broadcast address we have a pending PR: https://github.com/eclipse/smarthome/pull/4848

File: addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypes() {
+        return WizLightingBindingConstants.SUPPORTED_THING_TYPES_UIDS;
+    }
+
+    @Override
+    protected void startScan() {
+        if (inviteToken == null) {
+            String folderName = ConfigConstants.getUserDataFolder() + "/";
+            File file = new File(folderName, WizLightingBindingConstants.INVITE_TOKEN_TEMP_FILE);
```
[the remark this answers is not in the document]
Comment (basriram) [reply addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java]: @triller-telekom 

Thanks for this review and feedback, sorry took a while to get to this. I am incorporating most for the feedback but do have a question on this comment -

"I think this is a HACK and you should remove the file implementation. One can configure the binding via Paper UI and if one does not like to do that one can still use the generic configuration interface from ConfigAdmin.”

This one time use invitation code will be used to authenticate with the provider’s website to discover the bulbs that are configured for this home. Once that is done, this invitation code is useless. Hence the reason to use a temporary file to store the token for the very first time. Do you think there is a better way to handle this one time token?

[the remark this answers is not in the document]
Comment (triller-telekom) [reply addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java]: Without the token you are not able to "discover" new devices, right? I think the best way to do it is that you introduce a `bridge` thing which holds the token to access the devices on the web service. A user has to manually add a `bridge` from your binding and manually enter the token in its thing configuration. Afterwards you can start the scan on the bridge thing which then finds all devices, similar to the [Philips Hue binding](https://github.com/eclipse/smarthome/tree/master/extensions/binding/org.eclipse.smarthome.binding.hue), with the difference that you have the web service login as a bridge and the hue binding has a physical device which needs to be added before the discovery is able to find any bulbs.

**Comment by triller-telekom** [comment]
@basriram Are you still around? Would be a pity if we loose your contribution in this PR. I don't want to rush you, just curious if you are planning to continue working on this PR with he goal to get it merged.

If so, please create additional commits and do not amend your previous one, this makes a follow-up review a lot easier.

**Comment by triller-telekom** [comment]
@martinvw Can you please add an awaiting-feedback label on this PR?

**Comment by MaveryQ** [comment]
Sorry for my poor english, but i have a Wiz Light and i want to command it without the Wiz App on android.
So I would like to know what is the command line to send to the light trough http protocol ?
It seems like 
http://0123456789ab/192.168.0.17/39987/switch=ON      ?????

Thanks for your device.

**Comment by kaikreuzer** [comment]
@basriram Your feedback/response would be much appreciated (be it positive or negative)! Please let us know, whether you plan to finish this PR or not.

**Comment by triller-telekom** [comment]
@basriram As Kai mentioned in the previous post: Are you still planning to work on this binding? Please give us some feedback.

**Comment by MaveryQ** [comment]
Hello,

Any update ?

**Comment by basriram** [comment]
Hi,

I moved and don’t have access to the bulbs but let me try to clean up the code and submit it over the weekend.

Thanks

**Comment by MaveryQ** [comment]
the suspense is unsustainable ... :)

**Comment by alexsee** [comment]
I really would like to see the Wiz Lighting binding in OpenHAB. If there is some help needed here then I'm happy to volunteer :)

**Comment by J-N-K** [comment]
Closing due to inactivity for 9 month. If someone wants to continue, please clean-up the code and open a new PR.

**Comment by SRGDamia1** [comment]
I know this is old, but is there any chance of it being revisited?

@basriram, can you share more details of the API that you reverse engineered for the bulbs?

**Comment by basriram** [comment]
Yes absolutely. I would like to fix some open items but I don’t have those bulbs anymore. I will dig up my findings and send them over.

Thanks

File: addons/binding/org.openhab.binding.wizlighting/ESH-INF/thing/thing-types.xml [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        <item-type>Switch</item-type>
+        <label>Light Switch</label>
+        <description>Light Switch to turn on or off.</description>
+    </channel-type>
+    <channel-type id="color">
+        <item-type>Color</item-type>
+        <label>Color</label>
+        <category>ColorLight</category>
+    </channel-type>
+   <channel-type id="white">
```
Comment (SRGDamia1) [inline addons/binding/org.openhab.binding.wizlighting/ESH-INF/thing/thing-types.xml]: Cool white LEDs?

File: addons/binding/org.openhab.binding.wizlighting/ESH-INF/thing/thing-types.xml [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        <item-type>Number</item-type>
+        <label>White</label>
+    </channel-type> 
+   <channel-type id="white2">
```
Comment (SRGDamia1) [inline addons/binding/org.openhab.binding.wizlighting/ESH-INF/thing/thing-types.xml]: Warm white LED's?

File: addons/binding/org.openhab.binding.wizlighting/README.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+- WiZ Connected Lights Color A E26 Light Kit 
+
+
+## Discovery
+
+The devices are discovered only once when the app is installed for the first time. Once the app is configured it allows creating additional invite codes that allow installation of the app on other ios or android devices. In order to start auto discovery, create a file named wizlighting.token in Userdata folder with the 7 digit invite token and initiate discovery. If this invite code is provided, we mimic the flow of an ios device to obtain list of devices from the wiz world api end point. 
+
+## Binding Configuration
+
+The binding does not require any special configuration. The device should be connected to the same network.
```
[the remark this answers is not in the document]
Comment (SRGDamia1) [reply addons/binding/org.openhab.binding.wizlighting/README.md]: I *think* he might mean if you're using auto-discovery then you need the token to allow discovery but you don't need to configure anything else.  If you're manually configuration, you might need the mac addresses.

File: addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/WizLightingBindingConstants.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+     * List of all Thing Type UIDs.
+     */
+    public static final ThingTypeUID THING_TYPE_WIZ_BULB = new ThingTypeUID(BINDING_ID, "wizBulb");
+
+    /**
+     * List of all Channel ids
+     */
+    public static final String BULB_SWITCH_CHANNEL_ID = "switch";
+    public static final String BULB_COLOR_CHANNEL_ID = "color";
+    public static final String BULB_WHITE_CHANNEL_ID = "white";
```
Comment (SRGDamia1) [inline addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/WizLightingBindingConstants.java]: warm white

File: addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/WizLightingBindingConstants.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    public static final String BULB_C_CHANNEL_ID = "white2";
```
Comment (SRGDamia1) [inline addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/WizLightingBindingConstants.java]: cool white

File: addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/entities/SyncResponseParam.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+ * This POJO represents one Wiz Lighting Sync params
+ *
+ * @author Sriram Balakrishnan - Initial contribution
+ *
+ */
+public class SyncResponseParam extends Param {
+    public String mac;
+    public boolean state;
+    public int sceneId;
+    public boolean play;
```
Comment (SRGDamia1) [inline addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/entities/SyncResponseParam.java]: what is this?

**Comment by UselessMnemonic** [comment]
I am also interested in lending my hand.
I have absolutely no experience in reverse engineering an API. If someone could get on that, we can have a discovery service that is connected to the User's account.
Otherwise, the Home ID is enough to find lights through a broadcast channel. I think that's pretty fair.
I am currently extracting the protocol from this code, as well as from packet sniffing the Android App, for changing the state of the lights. It's very simple, nothing too exciting, just a handful of fields that describe exactly as they are named. I will attempt to show my findings later.
@SRGDamia1 What have you discovered so far? I see you are probing the white programs.

File: addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+    }
+
+    protected String obtainAccessToken()
+            throws URISyntaxException, InterruptedException, TimeoutException, ExecutionException {
+        String accessToken = null;
+        StringBuilder urlBuilder = new StringBuilder(WizLightingBindingConstants.DISCOVERY_API_OAUTH_URL);
+
+        logger.debug("Result {}", urlBuilder.toString());
+        Request request = httpClient.newRequest(urlBuilder.toString()).header("Authorization",
+                "Basic Ml81ZHE1cHFzNnI0b3dvNDAwb284MGNvd2d3azA0Y2drb2t3azB3bzgwa3drb2M4d2NjczpNV0ppWTJFeU1Ea3dNR0V3TVdJM1pqTTFOREppWm1RMU1qTm1NV05qTVRVNE9UZzFOag==")
```
[the remark this answers is not in the document]
Comment (SRGDamia1) [reply addons/binding/org.openhab.binding.wizlighting/src/main/java/org/openhab/binding/wizlighting/internal/discovery/WizLightingDiscoveryService.java]: This is the part I'm stumped by.  I think this was his specific bearer token from sign-in-with-google.  The WiZ app doesn't provide any ways of signing on except through Facebook and Google.

**Comment by SRGDamia1** [comment]
http://blog.dammitly.net/2019/10/cheap-hackable-wifi-light-bulbs-or-iot.html

**Comment by SRGDamia1** [comment]
I don't know java so I can't make the requested changes.  I'm also sort or baffled as to how to install it for myself to test.

Summary of things to do:
[ ] update copywrites
[ ] remove extra blank spaces
[ ] renames for clarity
[ ] change from nested ifs to switch/case where possible for clarity
[ ] add proper ds annotations and remove extra xmls
[ ] remove the definitions for finding own ip and mac addresses and use standard commands
[ ] combine switch/dimmer/color/white/white2 into a single channel color
[ ] figure out or completely remove the discovery service
    - To discover new devices you most *BOTH* log in to WiZ with either Google or Facebook and then after logged in, you use the simple token id from Google to go through oauth2 authentication with WiZ.  After you have the oauth token, you can use an invite token to get the list of devices.  I have no idea how to obtain that necessary first token from Google to even start using the invite token.
    - Within the local network, no authentication is needed - you can send commands and get responses directly from the bulbs.  So although it would be sad to loose the ability to get a list of bulbs in the house, you could force only manual thing addition and then not have to figure out the google/wiz combined authentication.

**Comment by alexsee** [comment]
I also have a Wiz blub, and did some modifications to the code. It's working pretty well so far. You can find it in my repository (https://github.com/alexsee/openhab2-addons/tree/wizlighting).
Since this work has been done a couple of month ago, I'm not totally aware what I exactly changed. Maybe it's worth looking at it.

The only anoying thing is that sometimes the state of the blub is not correctly updated when modifying it outside openHAB.

I can also contribute here. But first I need to get familiar with the new build system for openHAB 2.5. I guess we would also need to update the binding before handing it in again.

**Comment by schrieveslaach** [comment]
@alexsee, I updated your branch with master and created a new PR. I'm hoping to get this running with the help of you and others.

**Comment by SRGDamia1** [comment]
I've made a number of changes on my branch; did you look at those?

**Comment by schrieveslaach** [comment]
@SRGDamia1, I didn't notice that you made also some changes. Could you point out to me which changes you have made?

**Comment by SRGDamia1** [comment]
https://github.com/SRGDamia1/openhab2-addons/commits/master

I did some restructuring, added the required NotNullByDefault/Nullable annotations in all the places required, added the required OGSI context/reference annotations, removed the active discovery (because I couldn't figure out a way to obtain the required api token), added an RSSI channel, and added a bunch of comments.  I hadn't created my own PR yet because I've done a lot of fiddling while on the road and haven't actually tested all of the changes yet.

**Comment by schrieveslaach** [comment]
I looked at some of your changes and you have done a lot of work. Nice. :+1: 
There are many commits in your fork and I'm wondering what is the best approach to continue the development. 

Should I drop my PR and you will make a new one? Should we integrate your commits into my PR? (I would propose to rebase some commits)

What's your opinion here?

**Comment by alexsee** [comment]
@SRGDamia1 I finally could test your modifications. It works perfectly fine for me. I only found a bug relating to the `WizLightingHandler` where the PercentType cannot be converted to a `HSBType` (line 132). Not exactly sure how to handle that correctly.

**Comment by SRGDamia1** [comment]
@alexsee - great!

I'm still fiddling with a better way to try and grab the home id and configuration from the bulb, then I'll submit the new PR.

**Comment by SRGDamia1** [comment]
New PR!  https://github.com/openhab/openhab-addons/pull/6790

**Pull request closed without merging**